### PR TITLE
feat: add MongoDB Aggregation Pipeline support

### LINKED ISSUE #35: MongoDB Aggregation Pipeline Support
## Summary
Add support for MongoDB aggregation pipelines in the Kestra MongoDB plugin. Currently, the plugin only supports basic CRUD operations (Find, InsertOne, Update, Delete, Bulk, Load) but lacks support for MongoDB's aggregation framework.

## Motivation
MongoDB aggregation pipelines are essential for:
- Complex data transformations and analytics within the database
- Grouping, filtering, and calculating aggregated values (sum, average, count, etc.)
- Multi-stage data processing workflows
- Performance-optimized data analysis compared to multiple separate queries
- Advanced data manipulation, including joins, lookups, and reshaping

The aggregation pipeline is one of MongoDB's most powerful features, and its absence from the Kestra plugin significantly limits the plugin's utility.

### New Task: `Aggregate`

**Package**: `io.kestra.plugin.mongodb`
**Class**: `Aggregate`

**Key Properties**:
```yaml
- type: io.kestra.plugin.mongodb.Aggregate
  connection: # MongoDB connection details
  collection: "collection_name"
  pipeline: 
    - $match: { "status": "active" }
    - $group: 
        _id: "$category"
        total: { $sum: "$amount" }
        count: { $sum: 1 }
    - $sort: { "total": -1 }
  allowDiskUse: false  # Optional: allow stages to write to temp files
  maxTimeMs: 0         # Optional: maximum execution time
  batchSize: 1000      # Optional: cursor batch size
```

**Outputs**:
- `uri`: URI of the result file (similar to Find task)
- `size`: Number of documents returned
- `total`: Total number of documents processed

### Data Analytics
```yaml
- id: monthly_sales_report
  type: io.kestra.plugin.mongodb.Aggregate
  collection: orders
  pipeline:
    - $match: 
        createdAt: 
          $gte: "{{ trigger.date | dateAdd(-1, 'MONTHS') }}"
    - $group:
        _id: 
          year: { $year: "$createdAt" }
          month: { $month: "$createdAt" }
        totalRevenue: { $sum: "$amount" }
        orderCount: { $sum: 1 }
    - $sort: { "_id.year": 1, "_id.month": 1 }
```

### Data Transformation
```yaml
- id: user_profile_enrichment
  type: io.kestra.plugin.mongodb.Aggregate
  collection: users
  pipeline:
    - $lookup:
        from: "orders"
        localField: "_id"
        foreignField: "userId"
        as: "orderHistory"
    - $addFields:
        totalOrders: { $size: "$orderHistory" }
        avgOrderValue: { $avg: "$orderHistory.amount" }
    - $project:
        name: 1
        email: 1
        totalOrders: 1
        avgOrderValue: 1
```

### Real-time Metrics
```yaml
- id: real_time_dashboard_metrics
  type: io.kestra.plugin.mongodb.Aggregate  
  collection: events
  pipeline:
    - $match:
        timestamp: { $gte: "{{ trigger.date | dateAdd(-1, 'HOURS') }}" }
    - $group:
        _id: "$eventType"
        count: { $sum: 1 }
        latestTimestamp: { $max: "$timestamp" }
```

## Benefits

1. **Performance**: Aggregation pipelines execute within the database, reducing the number of runable tasks needed and data transformation within Kestra
2. **Flexibility**: Support for complex multi-stage data transformations
3. **Completeness**: Makes the MongoDB plugin feature-complete with core MongoDB functionality
4. **Integration**: Seamless integration with existing Kestra workflows and MongoDB tasks
5. **Scalability**: Built-in support for large dataset processing with `allowDiskUse` option

## Technical Considerations

- Leverage MongoDB Java Driver's `AggregateIterable` interface
- Support for all standard aggregation pipeline stages ($match, $group, $project, $sort, $lookup, etc.)
- Handle large result sets efficiently with streaming/batching
- Proper error handling and validation of pipeline syntax
- Support for aggregation options (allowDiskUse, maxTimeMs, etc.)

## Related MongoDB Documentation
- [Aggregation Pipeline](https://www.mongodb.com/docs/manual/core/aggregation-pipeline/)(https://www.mongodb.com/docs/manual/core/aggregation-pipeline/)
- [Aggregation Operations](https://www.mongodb.com/docs/manual/aggregation/)(https://www.mongodb.com/docs/manual/aggregation/)
- [Pipeline Stages](https://www.mongodb.com/docs/manual/reference/operator/aggregation-pipeline/)(https://www.mongodb.com/docs/manual/reference/operator/aggregation-pipeline/)

---

*This feature would greatly enhance Kestra's MongoDB integration and enable more sophisticated data processing workflows within the orchestration platform.*

### PR DESCRIPTION
**Pull request by RayedB**
Implements comprehensive MongoDB aggregation pipeline functionality to enable complex data transformations and analytics within Kestra workflows.
### What changes are being made and why?

  This PR implements comprehensive MongoDB Aggregation Pipeline support for the Kestra MongoDB plugin, enabling complex data transformations and analytics directly within MongoDB.

  **Key additions:**
  - New `Aggregate` task class with full aggregation pipeline support
  - Support for all MongoDB aggregation stages including `$match`, `$group`, `$project`, `$sort`, `$lookup`, `$unwind`, `$out`, `$merge`, etc.
  - Configurable options: `allowDiskUse`, `maxTimeMs`, `batchSize` for performance optimization
  - Support for both in-memory results and file storage (via `store` parameter)
  - Follows existing plugin patterns and conventions

  **Benefits:**
  - Enables complex data analytics and transformations directly in MongoDB
  - Reduces data movement between systems by processing within the database
  - Supports performance-optimized operations with `allowDiskUse` for large datasets
  - Provides feature parity with MongoDB's native aggregation capabilities

  Closes #35

  ---

  ### How the changes have been QAed?

The changes have been thoroughly tested with comprehensive integration tests covering various aggregation scenarios. Here are example flows demonstrating the functionality:

  **Example 1: Simple aggregation with grouping and sorting**
  ```yaml
  id: mongodb_aggregate_sales
  namespace: company.analytics

  tasks:
    - id: sales_by_category
      type: io.kestra.plugin.mongodb.Aggregate
      connection:
        uri: "mongodb://root:example@localhost:27017/?authSource=admin"
      database: "sales_db"
      collection: "transactions"
      pipeline:
        - $match:
            status: "completed"
        - $group:
            _id: "$category"
            totalAmount:
              $sum: "$amount"
            averageAmount:
              $avg: "$amount"
            transactionCount:
              $sum: 1
        - $sort:
            totalAmount: -1
        - $limit: 10

  Example 2: Complex aggregation with lookup and data transformation
  id: mongodb_user_analytics
  namespace: company.analytics

  tasks:
    - id: user_profile_enrichment
      type: io.kestra.plugin.mongodb.Aggregate
      connection:
        uri: "mongodb://root:example@localhost:27017/?authSource=admin"
      database: "analytics_db"
      collection: "users"
      pipeline:
        - $lookup:
            from: "orders"
            localField: "_id"
            foreignField: "userId"
            as: "userOrders"
        - $lookup:
            from: "reviews"
            localField: "_id"
            foreignField: "userId"
            as: "userReviews"
        - $addFields:
            totalOrders:
              $size: "$userOrders"
            totalSpent:
              $sum: "$userOrders.amount"
            avgOrderValue:
              $avg: "$userOrders.amount"
            reviewCount:
              $size: "$userReviews"
            avgRating:
              $avg: "$userReviews.rating"
        - $project:
            name: 1
            email: 1
            registrationDate: 1
            totalOrders: 1
            totalSpent: 1
            avgOrderValue: 1
            reviewCount: 1
            avgRating: 1
            customerSegment:
              $switch:
                branches:
                  - case:
                      $gte: ["$totalSpent", 1000]
                    then: "Premium"
                  - case:
                      $gte: ["$totalSpent", 500]
                    then: "Gold"
                  - case:
                      $gte: ["$totalSpent", 100]
                    then: "Silver"
                default: "Bronze"
        - $match:
            totalOrders:
              $gt: 0
      allowDiskUse: true
      maxTimeMs: 30000
      store: true

  Example 3: Time-based aggregation for reporting
  id: mongodb_monthly_report
  namespace: company.reports

  tasks:
    - id: monthly_revenue
      type: io.kestra.plugin.mongodb.Aggregate
      connection:
        uri: "mongodb://root:example@localhost:27017/?authSource=admin"
      database: "reporting_db"
      collection: "sales"
      pipeline:
        - $match:
            createdAt:
              $gte: "{{ trigger.date | dateAdd(-6, 'MONTHS') }}"
        - $group:
            _id:
              year:
                $year: "$createdAt"
              month:
                $month: "$createdAt"
            revenue:
              $sum: "$amount"
            orderCount:
              $sum: 1
            uniqueCustomers:
              $addToSet: "$customerId"
        - $addFields:
            uniqueCustomerCount:
              $size: "$uniqueCustomers"
        - $project:
            uniqueCustomers: 0
        - $sort:
            "_id.year": 1
            "_id.month": 1
      batchSize: 500

  Testing performed:
  - 8 comprehensive unit tests covering various aggregation scenarios
  - Integration tests with existing CRUD operations
  - Tests for string-based and map-based pipeline definitions
  - Validation of file storage option
  - Edge cases including empty pipelines and error handling

  ---
  Setup Instructions

  No special setup is required beyond the standard MongoDB connection. The aggregation functionality works with any MongoDB instance that supports aggregation pipelines (MongoDB 2.2+).

  - MongoDB connection URI with appropriate credentials
  - Database and collection access permissions
  - For large aggregations, ensure sufficient disk space when using allowDiskUse: true

### CODE REVIEW

File: src/main/java/io/kestra/plugin/mongodb/Aggregate.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        )
+    }
+)
+public class Aggregate extends AbstractTask implements RunnableTask<Aggregate.Output> {
+    @Schema(
+        title = "MongoDB aggregation pipeline.",
+        description = "List of pipeline stages as a BSON array or list of maps."
+    )
+    @PluginProperty(dynamic = true)
+    private List<Object> pipeline;
```
[the remark this answers is not in the document]
Comment (fdelbrayelle) [reply src/main/java/io/kestra/plugin/mongodb/Aggregate.java:113]: Plus you should add a `@NotNull` above `pipeline` since without it, the aggregration doesn't have any sense.

File: src/main/java/io/kestra/plugin/mongodb/Aggregate.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    )
+    @Builder.Default
+    private Property<Boolean> allowDiskUse = Property.ofValue(false);
+
+    @Schema(
+        title = "Maximum execution time in milliseconds.",
+        description = "Sets the maximum execution time on the server for this operation."
+    )
+    @Builder.Default
+    private Property<Integer> maxTimeMs = Property.ofValue(0);
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/mongodb/Aggregate.java:127]: Maybe increase the default value since if the user forget it the task will end as soon as it's run no?

File: src/test/java/io/kestra/plugin/mongodb/AggregateTest.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    private RunContextFactory runContextFactory;
+
+    @SuppressWarnings("unchecked")
+    @Test
+    void run() throws Exception {
+        RunContext runContext = runContextFactory.of(ImmutableMap.of());
+
+        Aggregate aggregate = Aggregate.builder()
+            .connection(MongoDbConnection.builder()
+                .uri(Property.ofValue("mongodb://root:example@localhost:27017/?authSource=admin"))
```
Comment (fdelbrayelle) [inline src/test/java/io/kestra/plugin/mongodb/AggregateTest.java:33]: The test cannot work without a MongoDB run thanks to testcontainers here or using a mock.

Comment (RayedB) [reply src/test/java/io/kestra/plugin/mongodb/AggregateTest.java]: went for TestContainers, to be confident of the behavior with a real mongodb instance

File: src/test/resources/flows/mongo-aggregate-example.yml [hunk 4]
```diff
@@ -0,0 +1,120 @@
+id: mongodb_aggregate_pipeline
+namespace: company.team
```
[the remark this answers is not in the document]
Comment (RayedB) [reply src/test/resources/flows/mongo-aggregate-example.yml:2]: added it in runner test

**Comment by RayedB** [comment]
@Malaydewangan09 i made a few changes based on your comments

**Review by Malaydewangan09 — APPROVED** [review]
LGTM, pushed last changes to fix the tests!
